Sources/Resolution/NamingServices/ENS.swift

```diff
@@ -0,0 +1,125 @@
+            throw ResolutionError.unsupportedNetwork
+        }
+        self.network = network
+        self.registryAddress = registryAddress
+        super.init(name: "ENS", providerUrl: providerUrl)
+    }
+
+    func isSupported(domain: String) -> Bool {
+        //Add additional domains
+        return domain.hasSuffix(".eth")
+    }
+
+    func owner(domain: String) throws -> String {
+        let tokenId = super.namehash(domain: domain)
+        guard let ownerAddress = try askRegistryContract(for: "owner", with: [tokenId]),
+            Utillities.isNotEmpty(ownerAddress) else {
+                throw ResolutionError.unregisteredDomain
+        }
+        return ownerAddress
+    }
+
+    func addr(domain: String, ticker: String) throws -> String {
+        let tokenId = super.namehash(domain: domain)
+        let key = "crypto.\(ticker.uppercased()).address"
+        let result = try record(tokenId: tokenId, key: key)
+        return result
+    }
+
+    // MARK: - Get Record
+    func record(domain: String, key: String) throws -> String {
+        let tokenId = super.namehash(domain: domain)
+        return try self.record(tokenId: tokenId, key: key)
+    }
+
+    func record(tokenId: String, key: String) throws -> String {
+        if key == "ipfs.html.value" {
+            let hash = try self.getContentHash(tokenId: tokenId)
+            return hash
+        }
+
+        let resolverAddress = try resolver(tokenId: tokenId)
+        let resolverContract = try super.buildContract(address: resolverAddress, type: .resolver)
+
+        let ensKeyName = self.fromUDNameToENS(record: key)
+
+        guard let result = try resolverContract.callMethod(methodName: "text", args: [tokenId, ensKeyName]) as? String,
+            Utillities.isNotEmpty(result) else {
+                throw ResolutionError.recordNotFound
+        }
+        return result
+    }
+
+    func records(keys: [String], for domain: String) throws -> [String: String] {
+        throw OtherError.runtimeError("Method not implemented.")
+    }
+
+    // MARK: - get Resolver
+    func resolver(domain: String) throws -> String {
+        let tokenId = super.namehash(domain: domain)
+        return try self.resolver(tokenId: tokenId)
+    }
+
+    func resolver(tokenId: String) throws -> String {
+        guard let resolverAddress = try askRegistryContract(for: "resolver", with: [tokenId]),
+            Utillities.isNotEmpty(resolverAddress) else {
+                throw ResolutionError.unconfiguredDomain
+        }
+        return resolverAddress
+    }
+
+    // MARK: - Helper functions
+    private func askRegistryContract(for methodName: String, with args: [String]) throws -> String? {
+        let registryContract: Contract = try super.buildContract(address: self.registryAddress, type: .registry)
+        return try registryContract.callMethod(methodName: methodName, args: args) as? String
+    }
+
+    private func fromUDNameToENS(record: String) -> String {
```

Comment:
I'd suggest following Microsoft guidelines for abbreviations: https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)?redirectedfrom=MSDN

```suggestion
    private func fromUDNameToEns(record: String) -> String {
```